system_zero/src/alu/rotate_shift.rs

```diff
@@ -0,0 +1,596 @@
+
+    // Output
+    let output_lo = lv[COL_ROTATE_SHIFT_OUTPUT_0];
+    let output_hi = lv[COL_ROTATE_SHIFT_OUTPUT_1];
+
+    // Check that every "bit" of exp_bits and top_bit is 0 or 1
+    exp_bits.map(|b| yield_constr.constraint(filter * (b * b - b)));
+    yield_constr.constraint(filter * (top_bit * top_bit - top_bit));
+
+    // Check that pow_exp = 2^exp, where exp is formed from the bits
+    // exp_bits[0..4].
+    //
+    // 2^exp = \prod_i=0^4 (2^(2^i) if exp_bits[i] = 1 else 1)
+    //       = \prod_i=0^4 ((2^(2^i) - 1) * exp_bits[i] + 1)
+    //       = pow_exp
+    //
+    // on the conditions that:
+    //
+    // pow_exp_aux_0 = \prod_i=0^1 ((2^i - 1) * exp_bits[i] + 1)
+    // pow_exp_aux_1 = \prod_i=2^3 ((2^i - 1) * exp_bits[i] + 1)
+    // pow_exp_aux_2 = pow_exp_aux_0 * pow_exp_aux_1
+    // pow_exp_mod32 = pow_exp_aux_2 * ((2^(2^4) - 1) * exp_bits[4] + 1)
+
+    let one = P::ONES;
+    // c[i-1] = 2^(2^i) - 1
+    let c = [1, 2, 3, 4].map(|i| P::from(F::from_canonical_u64(1u64 << (1u32 << i))) - P::ONES);
+
+    let constr1 = (exp_bits[0] + one) * (c[0] * exp_bits[1] + one);
+    yield_constr.constraint(filter * (constr1 - pow_exp_aux_0));
+    let constr2 = (c[1] * exp_bits[2] + one) * (c[2] * exp_bits[3] + one);
+    yield_constr.constraint(filter * (constr2 - pow_exp_aux_1));
+    let constr3 = pow_exp_aux_0 * pow_exp_aux_1;
+    yield_constr.constraint(filter * (constr3 - pow_exp_aux_2));
+    let constr4 = pow_exp_aux_2 * (c[3] * exp_bits[4] + one);
+    yield_constr.constraint(filter * (constr4 - pow_exp));
+
+    // An invalid shifted_lo (or _hi) can be too big to fit in
+    // Goldilocks field; e.g. if both _0 and _1 parts are 2^32-1, then
+    // shifted_lo = 2^32 - 1 + 2^32 (2^32 - 1) = 2^64 - 1 which
+    // overflows in GoldilocksField. Hence we check that
+    // shifted_{lo,hi} are valid Goldilocks elements following
+    // https://hackmd.io/NC-yRmmtRQSvToTHb96e8Q#Checking-element-validity
+    // The idea is check that a value v = (v_lo, v_hi) (32-bit words)
+    // satisfies the condition (v_lo == 0 OR v_hi != 2^32-1), which
+    // uses the structure of Goldilocks to check that v has the right
+    // form. The formula is:
+    //   v_lo * (one - aux * (u32_max - v_hi)) == 0
+    // where aux = (m32_max - v_hi)^-1 if it exists.
+
+    // u32_max = 2^32 - 1
+    let u32_max = P::from(F::from_canonical_u32(u32::MAX));
+
+    let constr1 = shifted_lo_aux_0 * (u32_max - shifted_lo_1);
+    yield_constr.constraint(filter * (constr1 - shifted_lo_aux_1));
+    let constr2 = shifted_hi_aux_0 * (u32_max - shifted_hi_1);
+    yield_constr.constraint(filter * (constr2 - shifted_hi_aux_1));
+
+    let shifted_lo_is_valid = shifted_lo_0 * (one - shifted_lo_aux_1);
+    let shifted_hi_is_valid = shifted_hi_0 * (one - shifted_hi_aux_1);
+
+    yield_constr.constraint(filter * shifted_lo_is_valid);
+    yield_constr.constraint(filter * shifted_hi_is_valid);
```

Comment:
Again, this checking validation of `shifted_lo_0` and `shifted_lo_1` (and `_hi`) could be done in a separate function just for clarity